lib/semian/protected_resource.rb

```diff
@@ -2,35 +2,59 @@ module Semian
   class ProtectedResource
     extend Forwardable
 
-    def_delegators :@resource, :destroy, :count, :semid, :tickets, :name
+    def_delegators :@bulkhead, :destroy, :count, :semid, :tickets, :name
     def_delegators :@circuit_breaker, :reset, :mark_failed, :mark_success, :request_allowed?,
                    :open?, :closed?, :half_open?
 
-    def initialize(resource, circuit_breaker)
-      @resource = resource
+    attr_reader :bulkhead, :circuit_breaker
+
+    def initialize(bulkhead, circuit_breaker)
+      @bulkhead = bulkhead
       @circuit_breaker = circuit_breaker
     end
 
     def destroy
-      @resource.destroy
-      @circuit_breaker.destroy
+      @bulkhead.destroy unless @bulkhead.nil?
+      @circuit_breaker.destroy unless @circuit_breaker.nil?
     end
 
     def acquire(timeout: nil, scope: nil, adapter: nil)
-      @circuit_breaker.acquire do
-        begin
-          @resource.acquire(timeout: timeout) do
-            Semian.notify(:success, self, scope, adapter)
+      if @circuit_breaker && !@bulkhead
+        acquire_circuit_breaker(scope, adapter) do
+          yield self
+        end
+      elsif @bulkhead && !@circuit_breaker
+        acquire_bulkhead(timeout, scope, adapter) do
+          yield self
+        end
+      else
+        acquire_circuit_breaker(scope, adapter) do
+          acquire_bulkhead(timeout, scope, adapter) do
             yield self
           end
-        rescue ::Semian::TimeoutError
-          Semian.notify(:busy, self, scope, adapter)
-          raise
         end
       end
+    end
```

Comment:
If you return `yield` in `acquire_*` if it's not enabled, you might be able to do something like this:

```ruby
def acquire(timeout: nil, scope: nil, adapter: nil, &block)
  [method(:acquire_bulkhead), method(:acquire_circuit_breaker)].inject(block) { |nested, method| 
    ->(scope, adapter) { method.call(scope, adapter, &nested) } 
  }.call(scope, adapter)
end 
```

The idea is to build a proc that properly nests the calls, with bulkhead first (note it'll build it reverse from what you might expect).